include/api/CDataFrameAnalysisSpecification.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_api_CDataFrameAnalysisSpecification_h
+#define INCLUDED_ml_api_CDataFrameAnalysisSpecification_h
+
+#include <core/CFastMutex.h>
+
+#include <api/ImportExport.h>
+
+#include <rapidjson/fwd.h>
+
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace ml {
+namespace core {
+class CDataFrame;
+}
+namespace api {
+class CDataFrameAnalysisRunner;
+class CDataFrameAnalysisRunnerFactory;
+
+//! \brief Parses a complete specification for running a core::CDataFrame analysis
+//! and supports launching that analysis on a specified frame object.
+//!
+//! DESCRIPTION:\n
+//! This manages extracting all configuration for a particular analysis from a JSON
+//! header which is passed to the data_frame_analyzer command before any data. This
+//! creates and owns an analysis runner object which is also configured by the header.
+//! The analysis is run asynchronously via the CDataFrameAnalysisSpecification::run
+//! method which returns a handle to the runner to retrieve progress and errors.
+class API_EXPORT CDataFrameAnalysisSpecification {
+public:
+    using TStrVec = std::vector<std::string>;
+    using TRunnerPtr = std::unique_ptr<CDataFrameAnalysisRunner>;
+    using TRunnerFactoryPtr = std::unique_ptr<CDataFrameAnalysisRunnerFactory>;
```

Comment:
There are many places in the code where the shortened form `UPtr` has been used for `unique_ptr`.  I'm not sure if this is consistent at the moment but it might be nice to standardise on `Ptr` for shared pointers and `UPtr` for unique pointers (unless we're so far from consistency at the moment that that would be futile).

Reply:
Agreed.

Reply:
[e6e7868](https://github.com/elastic/ml-cpp/pull/306/commits/e6e7868318e3b332fd9dcb92f786aed30255a02e)